### ISSUE TITLE
Which ImageFormat to use.

### ISSUE THREAD

**Issue by elblogbruno** [homuler/MediaPipeUnityPlugin#43]
Hello!
As I told you on error #23 I am integrating ar foundation into this. You told me to create a ImageFrame, the thing is I don't know which Format Should be use,
Glad if you or someone could help me out
Cordially
Bruno

**Comment by elblogbruno** [comment]
also can't seem to push and ImageFrame to a graph
![image](https://user-images.githubusercontent.com/10481058/106027249-009f4e80-60cb-11eb-9d46-a69cf7d40855.png)

**Comment by homuler** [comment]
Of course it depends on how the `buffer` (`NativeArray<byte>`) is aligned.
But MediaPipe does not support all the format for all the calculations (e.g. you cannot convert SBGRA to GpuBuffer https://github.com/google/mediapipe/blob/master/mediapipe/gpu/gpu_buffer_format.cc#L176), and in such a case, this plugin will throw an error or aborts in the worst case.

> also can't seem to push and ImageFrame to a graph

Probably you are using `DemoGraph` and `DemoGraph` receives `TextureFrame`, not `ImageFrame`.

By the way, I recommend you not using `DemoGraph` or scripts under `Examples` currently.
Most of them are implemented only for the demonstration purpose (though some useful classes may be moved under `SDK` in the future).

**Comment by homuler** [comment]
If more generalized (platform independent) graph implementation is found, it will be placed under `SDK`, but not yet.

**Comment by elblogbruno** [comment]
So I can simply pass an ImageFrame like in the picture? I have seen TextureFrame has some callbacks and things.

**Comment by elblogbruno** [comment]
![image](https://user-images.githubusercontent.com/10481058/106134793-532e4880-6167-11eb-93de-18390c2e1233.png)
hmm...

**Comment by elblogbruno** [comment]
i am looking at the logs

**Comment by homuler** [comment]
> I have seen TextureFrame has some callbacks and things.

If you pass an OpenGL ES texture directly to MediaPipe, those callbacks are necessary, but as long as you pass `ImageFrame`s to MediaPipe, they are unnecessary.

**Comment by elblogbruno** [comment]
thanks
![image](https://user-images.githubusercontent.com/10481058/106137053-760e2c00-616a-11eb-9b5d-373ef73829ac.png)
just for trying to get it work I am modifying the DemoGraph

**Comment by Codeavr** [comment]
If you are wishing to integrate ARFoundation, consider that `Convert` function returns byte array which you can directly pass to the input stream. No need in glcontext, callbacks and etc. Works good for me

**Comment by elblogbruno** [comment]
Yes exactly @Codeavr I do get a vyte array but don't know how to pass it to the input stream!
Could you shine a light on me please? Thanks

**Comment by elblogbruno** [comment]
![image](https://user-images.githubusercontent.com/10481058/106161104-d9a65280-6186-11eb-93b7-390f263791ec.png)
i was trying to create an ImageFrame out of it but I can't get it to work

**Comment by Codeavr** [comment]
@elblogbruno  Well, in my project I have quite different pipeline. 
I need converted video texture preview, so I load ARFoundation texture to Texture2D
```
var buffer = new NativeArray<byte>(size, Allocator.Temp);

//unsafe context
image.Convert(conversionParams, new IntPtr(buffer.GetUnsafePtr()), buffer.Length);

_texture.LoadRawTextureData(buffer);
_texture.Apply();
```
then I just  image frame
```
imageFrame = new ImageFrame
(
    ImageFormat.Format.SRGBA,
    _texture.width,
    _texture.height,
    4 * _texture.width,
    _texture.GetRawTextureData<byte>()
);
```

**Comment by elblogbruno** [comment]
Hi @Codeavr thanks for answering. How do you pass that imageFrame to the graph?

**Comment by Codeavr** [comment]
@elblogbruno Well, in my project I have quite different pipeline.
I need converted video texture preview, so I load ARFoundation texture to Texture2D
```c#
//unsafe context
image.Convert(conversionParams, new IntPtr(buffer.GetUnsafePtr()), buffer.Length);

_texture.LoadRawTextureData(buffer);
_texture.Apply();
```
then I create `ImageFrame` using this texture pointer to data and pass it to the input stream
```c#
imageFrame = new ImageFrame
(
    ImageFormat.Format.SRGBA,
    _texture.width,
    _texture.height,
    4 * _texture.width,
    _texture.GetRawTextureData<byte>()
);

var packet = new ImageFramePacket(imageFrame, timestamp);

_graph.AddPacketToInputStream(InputStream, packet);
```

There must be more performing way to this, but this save/load overhead is not that high

**Comment by elblogbruno** [comment]
ooh thanks!

**Comment by elblogbruno** [comment]
but are you planning to work on Android?

**Comment by Codeavr** [comment]
> but are you planning to work on Android? I believe for android more things should be done, does this current setup work on android for you?
I started working with this plugin before homuler implemented async gpu texture loading (those things with texture pooling and callbacks), and I'm still with old code on android which look kinda: 
```c#
var status = _gpuHelper.RunInGlContext(() => {
    var texture = _gpuHelper.CreateSourceTexture(imageFrame);
    var gpuFrame = texture.GetGpuBufferFrame();
    texture.Release();

    return _graph.AddPacketToInputStream(InputStream, new GpuBufferPacket(gpuFrame, timestamp));
});
```
I'm not sure if it's completely correct, but works for me

**Comment by elblogbruno** [comment]
hhm thanks!

**Comment by elblogbruno** [comment]
@Codeavr how you are rendering the output?

**Comment by Codeavr** [comment]
@elblogbruno if you are about annotations, I use orthographic projection on my camera and just drawing landmarks into linerenderers

**Comment by elblogbruno** [comment]
Well I was speaking about graph.renderoutput() function. I will make the annotations myself in the near future , I still need to get image frame passing working.
I can't seem to be able.
Thanks @Codeavr

**Comment by homuler** [comment]
> ```cs
> var status = _gpuHelper.RunInGlContext(() => {
>     var texture = _gpuHelper.CreateSourceTexture(imageFrame);
>     var gpuFrame = texture.GetGpuBufferFrame();
>     texture.Release();
>
>    return _graph.AddPacketToInputStream(InputStream, new GpuBufferPacket(gpuFrame, timestamp));
>  });
> ```

It's correct, but using `ImageFrameToGpuBufferCalculator`, you can simply call `AddPacketToInputStream` with `ImageFramePacket` and it's faster (`ImageFrameToGpuBufferCalculator` does the same thing in native code).

https://github.com/homuler/MediaPipeUnityPlugin/blob/master/Assets/MediaPipe/Examples/Graphs/FaceDetection/Resources/face_detection_desktop_gpu.txt#L51-L55

```cs
using (var packet = new ImageFramePacket(imageFrame, timestamp)) {
  var status = _graph.AddPacketToInputStream(InputStream, packet);
}
```

>  I believe for android more things should be done, does this current setup work on android for you?

On Android, you can pass texture pointers to MediaPipe, and the setup becomes complicated if you'd like to do so, but you don't have to do so.
Desktop GPU samples should work on Android too (if `UNITY_ADNROID` pragma is removed. Probably I should have used custom directives).

**Comment by homuler** [comment]
> I was speaking about graph.renderoutput() function.

Except for `OfficialDemoAndroid`, `DemoGraph#RenderOutput` is not platform dependent.

**Comment by Codeavr** [comment]
@elblogbruno I don't understand the question

**Comment by elblogbruno** [comment]
> On Android, you can pass texture pointers to MediaPipe, and the setup becomes complicated if you'd like to do so, but you don't have to do so.

But passing the pointer is a method that improves the performance right?

> Desktop GPU samples should work on Android too (if `UNITY_ADNROID` pragma is removed. Probably I should have used custom directives).

I will continue today trying to get it working on Android! Thanks for the support and tips

**Comment by elblogbruno** [comment]
> @elblogbruno I don't understand the question

Nevermind it was a nonsense question I answered myself 😂

**Comment by elblogbruno** [comment]
![image](https://user-images.githubusercontent.com/10481058/106271447-d61ad600-622f-11eb-941b-f3df7bb1690c.png)
What could this be?

**Comment by homuler** [comment]
Please uncomment this line, then build and rerun. 
https://github.com/homuler/MediaPipeUnityPlugin/blob/master/C/WORKSPACE#L48

Glog will output the error log.
Or if you can, set `GLOG_v=2` and run it.

**Comment by elblogbruno** [comment]
ok thanks! I will debug and return here with an answer!

**Comment by elblogbruno** [comment]
`Log file created at: 2021/01/29 13:34:57
Running on machine: localhost
Log line format: [IWEF]yyyymmdd hh:mm:ss.uuuuuu threadid file:line] msg
W20210129 13:34:57.718008 27261 calculator_graph.cc:939] INVALID_ARGUMENT: Packet type mismatch on calculator outputting to stream "input_video": The Packet stores "mediapipe::ImageFrame", but "mediapipe::GpuBuffer" was requested.`

**Comment by elblogbruno** [comment]
what could this be?

**Comment by homuler** [comment]
Maybe you are pushing `ImageFramePacket`s to `input_video`, but `input_video` should be defined as `GpuBuffer`.
Will you share the config you use?

**Comment by elblogbruno** [comment]
I am using the hand_tracking_android.txt config.
https://pastebin.com/izyJ48J6
Here is what I have modified from DemoGraph.cs
https://pastebin.com/6UUSz9QJ

And here is the WebcamScreenController.cs modified with arFoundation and pushing imageFrame to modified DemoGraph.cs
https://pastebin.com/rhiS8SQr

**Comment by elblogbruno** [comment]
I did switched from outputTexture to the buffer byte array, and now I get it to show but with a poor,poor performance

**Comment by homuler** [comment]
Stop rendering annotations temporarily, and try `ObserveOutputStream` as described here https://github.com/homuler/MediaPipeUnityPlugin/issues/23#issuecomment-763247582 .
In this case, you can remove `PacketPresenceCalculator`s from the config.

You cannot get output packets faster than this way, so I recommend you check the maximum performance first.
(To be precise, if you prepare specialized APIs to send image data, I think you can improve the performance, but I don't know how much it will be.)

**Comment by elblogbruno** [comment]
hmm ok thanks!

**Comment by elblogbruno** [comment]
and can I pass an ImageFrame to the OfficialGraph?
sorry for asking lot of questions recently

**Comment by homuler** [comment]
Yes, but note that you need to convert `ImageFrame` to `GpuBuffer` (https://github.com/homuler/MediaPipeUnityPlugin/issues/43#issuecomment-769498179) if you use `official_demo_android.txt`.
`OfficialDemoGpu` would also work, but this graph does not render the output image directly to `Texture2D`.

**Comment by kitmek1** [comment]
C:\Users\kitme\OneDrive\Desktop\New folder (2)\MediaPipeUnityPlugin>make cpu
./.build/protobuf-3.13.0/csharp/buildall.sh
process_begin: CreateProcess(NULL, bash C:\Users\kitme\OneDrive\Desktop\New folder (2)\MediaPipeUnityPlugin\.build\protobuf-3.13.0\csharp\buildall.sh, ...) failed.
make (e=2): The system cannot find the file specified.
Makefile:61: recipe for target '.build/protobuf-3.13.0/csharp/src/Google.Protobuf/bin/Release/net45/Google.Protobuf.dll' failed
mingw32-make.exe: *** [.build/protobuf-3.13.0/csharp/src/Google.Protobuf/bin/Release/net45/Google.Protobuf.dll] Error 2
 i have this error while loading model can you help me...

**Comment by elblogbruno** [comment]
> Yes, but note that you need to convert `ImageFrame` to `GpuBuffer` ([#43 (comment)](https://github.com/homuler/MediaPipeUnityPlugin/issues/43#issuecomment-769498179)) if you use `official_demo_android.txt`.
> `OfficialDemoGpu` would also work, but this graph does not render the output image directly to `Texture2D`.

Thanks!

**Comment by elblogbruno** [comment]
![image](https://user-images.githubusercontent.com/10481058/106389211-7ca6d880-63e2-11eb-8cda-72a20371026f.png)
What could this be?

**Comment by elblogbruno** [comment]
> C:\Users\kitme\OneDrive\Desktop\New folder (2)\MediaPipeUnityPlugin>make cpu
> ./.build/protobuf-3.13.0/csharp/buildall.sh
> process_begin: CreateProcess(NULL, bash C:\Users\kitme\OneDrive\Desktop\New folder (2)\MediaPipeUnityPlugin.build\protobuf-3.13.0\csharp\buildall.sh, ...) failed.
> make (e=2): The system cannot find the file specified.
> Makefile:61: recipe for target '.build/protobuf-3.13.0/csharp/src/Google.Protobuf/bin/Release/net45/Google.Protobuf.dll' failed
> mingw32-make.exe: *** [.build/protobuf-3.13.0/csharp/src/Google.Protobuf/bin/Release/net45/Google.Protobuf.dll] Error 2
> i have this error while loading model can you help me...

make sure you have installed protobuf, but don't post your issue on a thread about a non-related issue to your issue.

**Comment by homuler** [comment]
TL;DR
Call `RunInGlContext(NativeGlStatusFunction)` instead or use Mono as the scripting backend.

`GlStatusFunction` returns `Status`, but MediaPipe (C++) doesn't understand `Status` (C# object), so when `RunInGlContext(GlStatusFunction` is called, this plugin converts the `GlStatusFunction` to an anonymous `NativeGlStatusFunction` (returns a native pointer).
https://github.com/homuler/MediaPipeUnityPlugin/blob/09cd496124811a3e2392bafbb07f6825dfc2988d/Assets/MediaPipe/SDK/Scripts/Gpu/GlCalculatorHelper.cs#L37-L56

However, IL2CPP does not support marshaling non-static method to native code, and Unity thorws the above error if we try to do that or pin function pointers to instance methods (the latter in the above case).
To avoid this error, define a static `NativeGlStatusFunction` and call `RunInGlContext(NativeGlStatusFunction)` with it instead.

See also ↓.
https://github.com/homuler/MediaPipeUnityPlugin/blob/09cd496124811a3e2392bafbb07f6825dfc2988d/Assets/MediaPipe/Examples/Graphs/OfficialDemo/Scripts/OfficialDemoAndroid.cs#L56-L67

**Comment by elblogbruno** [comment]
I will try that! Thanks for all the help 😊

**Comment by elblogbruno** [comment]
I am still getting the same issue wth.

**Comment by homuler** [comment]
Will you share the related code? I cannot explain more clearly than https://github.com/homuler/MediaPipeUnityPlugin/issues/43#issuecomment-770498550 .

By the way, it'd be much better to copy code, logs, etc... in text format than in image (e.g. screenshot).
It's difficult for us to copy texts in image or control the size of them, and those texts won't be indexed by search engine.

**Comment by elblogbruno** [comment]
Sure, here is my modified code https://pastebin.com/tfY5tX7u using  #43 (comment).
Also here is the code that creates and pushes the ImageFrame https://pastebin.com/8FeaWH6E
Sorry about sending pictures before it was a quick and syntax highlighted way of sharing.
Hope you see what I am doing wrong.
Thanks

**Comment by homuler** [comment]
`NativeGlStatusFunction` should return `IntPtr`, not `Status`.

```cs
  [AOT.MonoPInvokeCallback(typeof(GlCalculatorHelper.NativeGlStatusFunction))]
  static IntPtr GetStatusAndroid() {
    var texture = gpuHelper.CreateSourceTexture(currentTextureFrame);
    var gpuFrame = texture.GetGpuBufferFrame();
    texture.Release();
  Debug.Log("Getting status android");
    return graph.AddPacketToInputStream(inputStream, new GpuBufferPacket(gpuFrame, currentTimestamp)).mpPtr;
  }
```

Note that the above code may throw exceptions, and C++ cannot handle them.
e.g. https://github.com/homuler/MediaPipeUnityPlugin/blob/09cd496124811a3e2392bafbb07f6825dfc2988d/Assets/MediaPipe/Examples/Scripts/DemoGraph.cs#L69-L82

**Comment by elblogbruno** [comment]
Thanks for the help!

**Comment by elblogbruno** [comment]
I get it working but can't read from the stream consistently. Hand annotations are shown as always but as you see by the code, i am debugging some values. 

` public override void RenderOutput(WebCamScreenController screenController, ImageFrame textureFrame) {
    var handTrackingValue = FetchNextHandTrackingValue();
    RenderAnnotation(screenController, handTrackingValue);
    Debug.Log(handTrackingValue.PalmDetections.Count);
    Debug.Log(handTrackingValue.Handednesses[0]);
    screenController.DrawScreen(textureFrame);
  }
`

Values are debugged for some seconds until it stops with the unexpected END OF FILE issue. Even though values are not being printed out on the console annotations still appear on screen. I will also try now to debug landmarks on the official graph.

`
2021/02/02 12:01:53.617 21374 21908 Info native I20210202 12:01:53.617295 21908 scheduler.cc:455] active queues: 1
2021/02/02 12:01:53.617 21374 21904 Info native I20210202 12:01:53.617719 21904 calculator_node.cc:812] Calling Calculator::Process() for node: handlandmarktrackinggpu__handlandmarkgpu__ImageToTensorCalculator timestamp: 36
2021/02/02 12:01:53.618 21374 21904 Info native I20210202 12:01:53.618667 21904 calculator_node.cc:825] Called Calculator::Process() for node: handlandmarktrackinggpu__handlandmarkgpu__ImageToTensorCalculator timestamp: 36
2021/02/02 12:01:53.618 21374 21904 Info native I20210202 12:01:53.618875 21904 calculator_node.cc:812] Calling Calculator::Process() for node: handlandmarktrackinggpu__handlandmarkgpu__InferenceCalculator timestamp: 36
0001/01/01 00:00:00.000 -1 -1 Info  read: unexpected EOF!
`
Again thanks for your help @homuler !

**Comment by homuler** [comment]
Will you set `GLOG_v=5` and share the full log, please?
cf. https://github.com/homuler/MediaPipeUnityPlugin/blob/master/Assets/MediaPipe/Examples/Scripts/SceneDirector.cs#L32

**Comment by elblogbruno** [comment]
https://pastebin.com/NhJ8n5UT
As you seeit shows the message DEBUGGING VALUES FROM STREAM and prints out handTrackingValue.PalmDetections[0] and handTrackingValue.Handednesses[0] once. and it does not show anymore.

**Comment by elblogbruno** [comment]
`Log file created at: 2021/02/02 11:43:24
Running on machine: localhost
Log line format: [IWEF]yyyymmdd hh:mm:ss.uuuuuu threadid file:line] msg
F20210202 11:43:24.849306 16002 calculator_graph.cc:124] Non-OK-status: Initialize(config) status: INVALID_ARGUMENT: while processing the input side packets of subgraph node HandRendererSubgraph: ; Tag "DESTINATION" does not exist in the subgraph config.
E20210202 11:43:24.854162 16002 calculator_graph.cc:36] Aborted
`
Glog log

**Comment by homuler** [comment]
> Error Unity ArgumentOutOfRangeException: Index was out of range. Must be non-negative and less than the size of the collection.

I think simply your code has some bugs.

```cs
Debug.Log(handTrackingValue.Handednesses[0]);
```

**Comment by elblogbruno** [comment]
Ok I will investigate furthermore!
and why I can't fetch next packet from OfficalDemoGraph?
`public override void RenderOutput(WebCamScreenController screenController, ImageFrame textureFrame) {
    // MediaPipe renders to the texture directly.
    //var landmarks = FetchNext(handLandmarksStreamPoller, handLandmarksPacket, "hand_landmarks");
    //Debug.Log("render image frame: " + landmarks);
    return;
  }`
I have uncommented it and texture is displayed,  but when I call the function FetchNext it does not work. I followed #23 tip from your part on how to read values synchronously.
thanks

**Comment by homuler** [comment]
Did you read this comment?
```cs
// Note that the above method blocks the thread, and if the output packet is empty, it never returns (the process hangs).
// see https://github.com/homuler/MediaPipeUnityPlugin/blob/master/Assets/MediaPipe/Examples/Graphs/HandTracking/Scripts/HandTrackingGraph.cs
```

Before calling `FetchNext`, you have to make sure that the output packet exists.
If you use `ObserveOutputStream`, this problem won't occur.

**Comment by elblogbruno** [comment]
oops, I am so sorry I missed it! Thanks

**Comment by elblogbruno** [comment]
I was implementing `ObserveOutputStream` and always get this issue, that has popped up before. It is because   `private Status OutputCallback(NormalizedLandmarkListVectorPacket packet)` should be static or decorator   [AOT.MonoPInvokeCallback(typeof(GlCalculatorHelper.NativeGlStatusFunction))] needs to be added?

[https://github.com/homuler/MediaPipeUnityPlugin/blob/master/Assets/MediaPipe/Examples/Graphs/OfficialDemo/Scripts/OfficialDemoDesktop.cs](url)
I have followed this as a reference!


```
2021/02/02 16:20:22.948 22850 22893 Error Unity ArgumentException: Object contains non-primitive or non-blittable data.
2021/02/02 16:20:22.948 22850 22893 Error Unity   at System.Runtime.InteropServices.GCHandle..ctor (System.Object value, System.Runtime.InteropServices.GCHandleType type) [0x00000] in <00000000000000000000000000000000>:0 
2021/02/02 16:20:22.948 22850 22893 Error Unity   at System.Runtime.InteropServices.GCHandle.Alloc (System.Object value, System.Runtime.InteropServices.GCHandleType type) [0x00000] in <00000000000000000000000000000000>:0 
2021/02/02 16:20:22.948 22850 22893 Error Unity   at Mediapipe.CalculatorGraph.ObserveOutputStream[T,U] (System.String streamName, Mediapipe.CalculatorGraph+PacketCallback`2[T,U] packetCallback, System.Runtime.InteropServices.GCHandle& callbackHandle) [0x00000] in <00000000000000000000000000000000>:0 
2021/02/02 16:20:22.948 22850 22893 Error Unity   at OfficialDemoAndroid.StartRun (UnityEngine.Texture texture) [0x00000] in <00000000000000000000000000000000>:0 
2021/02/02 16:20:22.948 22850 22893 Error Unity   at WebCamScreenController+<RunGraph>d__45.MoveNext () [0x00000] in <00000000000000000000000000000000>:0 
2021/02/02 16:20:22.948 22850 22893 Error Unity   at UnityEngine.SetupCoroutine.InvokeMoveNext (System.Collections.IEnumerator enumerator, System.IntPtr returnValueAddress) [
2021/02/02 16:20:24.695 22850 23180 Error native motion_tracking_context.cc:1064 Unable to find measurements corresponding to VIO status at timestamp 2568738454048105.

```

**Comment by homuler** [comment]
The callback type is `CalculatorGraph.NativePacketCallback` and it should return `IntPtr`, not `Status` (the reason is same as https://github.com/homuler/MediaPipeUnityPlugin/issues/43#issuecomment-770498550).
See also https://github.com/homuler/MediaPipeUnityPlugin/blob/master/Assets/MediaPipe/SDK/Scripts/Framework/CalculatorGraph.cs#L59 .

```cs
[AOT.MonoPInvokeCallback(typeof(CalculatorGraph.NativePacketCallback))]
static HandLandmarksCallback(IntPtr ptr) {
  using (var packet = new NormalizedLandmarksListVectorPacket(ptr, false)) {
    using (var landmarks = packet.Get()) {
      // do something
    }
    return Status.Ok().mpPtr;
  }
}
```

**Comment by elblogbruno** [comment]
Ok I see what was the problem!
Thanks for all again!

**Comment by homuler** [comment]
I'm sorry that `ObserveOutputStream` is not documented well and no examples exist for IL2CPP, but note that the packet the callback receives is still owned by MediaPipe, so the constructor's second argument has to be `false`.

```cs
var packet = new NormalizedLandmarksListVectorPacket(ptr, false);
```

**Comment by elblogbruno** [comment]
Don't worry about having less documentation. You are doing a great job with this repo!

**Comment by elblogbruno** [comment]
![image](https://user-images.githubusercontent.com/10481058/106626262-cd036f00-6577-11eb-87c4-75d0adf97b57.png)
it expects an status

**Comment by elblogbruno** [comment]
> I got it working 😄 but can't read from the stream consistently. Hand annotations are shown as always but as you see by the code, i am debugging some values.
> 
> `public override void RenderOutput(WebCamScreenController screenController, ImageFrame textureFrame) { var handTrackingValue = FetchNextHandTrackingValue(); RenderAnnotation(screenController, handTrackingValue); Debug.Log(handTrackingValue.PalmDetections.Count); Debug.Log(handTrackingValue.Handednesses[0]); screenController.DrawScreen(textureFrame); }`
> 
> Values are debugged for some seconds until it stops with the unexpected END OF FILE issue. Even though values are not being printed out on the console annotations still appear on screen. I will also try now to debug landmarks on the official graph. (i am getting better performance with official graph, but can't print out values on console)
> 
> `2021/02/02 12:01:53.617 21374 21908 Info native I20210202 12:01:53.617295 21908 scheduler.cc:455] active queues: 1 2021/02/02 12:01:53.617 21374 21904 Info native I20210202 12:01:53.617719 21904 calculator_node.cc:812] Calling Calculator::Process() for node: handlandmarktrackinggpu__handlandmarkgpu__ImageToTensorCalculator timestamp: 36 2021/02/02 12:01:53.618 21374 21904 Info native I20210202 12:01:53.618667 21904 calculator_node.cc:825] Called Calculator::Process() for node: handlandmarktrackinggpu__handlandmarkgpu__ImageToTensorCalculator timestamp: 36 2021/02/02 12:01:53.618 21374 21904 Info native I20210202 12:01:53.618875 21904 calculator_node.cc:812] Calling Calculator::Process() for node: handlandmarktrackinggpu__handlandmarkgpu__InferenceCalculator timestamp: 36 0001/01/01 00:00:00.000 -1 -1 Info read: unexpected EOF!`
> Again thanks for your help @homuler !

getting this problem again wth

**Comment by homuler** [comment]
Please see the link in the https://github.com/homuler/MediaPipeUnityPlugin/issues/43#issuecomment-771730648 .
`ObserveOutputStream` is overloaded.

**Comment by elblogbruno** [comment]
I am using this version of the ObserveOutputStream https://github.com/homuler/MediaPipeUnityPlugin/blob/master/Assets/MediaPipe/SDK/Scripts/Framework/CalculatorGraph.cs#L59
As the other one expects a status in the static function callback

**Comment by elblogbruno** [comment]
> ![image](https://user-images.githubusercontent.com/10481058/106626262-cd036f00-6577-11eb-87c4-75d0adf97b57.png)
> it expects an status 

See this

**Comment by homuler** [comment]
It's because the arity is different...
```cs
graph.ObserveOutputStream("hand_landmarks", HandLandmarksCallback).AssertOk();
```

**Comment by elblogbruno** [comment]
> It's because the arity is different...
> ```cs
> graph.ObserveOutputStream("hand_landmarks", HandLandmarksCallback).AssertOk();
> ```

Yes I am using it like this

**Comment by elblogbruno** [comment]
> > I got it working 😄 but can't read from the stream consistently. Hand annotations are shown as always but as you see by the code, i am debugging some values.
> > 
> > `public override void RenderOutput(WebCamScreenController screenController, ImageFrame textureFrame) { var handTrackingValue = FetchNextHandTrackingValue(); RenderAnnotation(screenController, handTrackingValue); Debug.Log(handTrackingValue.PalmDetections.Count); Debug.Log(handTrackingValue.Handednesses[0]); screenController.DrawScreen(textureFrame); }`
> > 
> > Values are debugged for some seconds until it stops with the unexpected END OF FILE issue. Even though values are not being printed out on the console annotations still appear on screen. I will also try now to debug landmarks on the official graph. (i am getting better performance with official graph, but can't print out values on console)
> > 
> > `2021/02/02 12:01:53.617 21374 21908 Info native I20210202 12:01:53.617295 21908 scheduler.cc:455] active queues: 1 2021/02/02 12:01:53.617 21374 21904 Info native I20210202 12:01:53.617719 21904 calculator_node.cc:812] Calling Calculator::Process() for node: handlandmarktrackinggpu__handlandmarkgpu__ImageToTensorCalculator timestamp: 36 2021/02/02 12:01:53.618 21374 21904 Info native I20210202 12:01:53.618667 21904 calculator_node.cc:825] Called Calculator::Process() for node: handlandmarktrackinggpu__handlandmarkgpu__ImageToTensorCalculator timestamp: 36 2021/02/02 12:01:53.618 21374 21904 Info native I20210202 12:01:53.618875 21904 calculator_node.cc:812] Calling Calculator::Process() for node: handlandmarktrackinggpu__handlandmarkgpu__InferenceCalculator timestamp: 36 0001/01/01 00:00:00.000 -1 -1 Info read: unexpected EOF!`
> > Again thanks for your help @homuler !
> 
> getting this problem again wth

And getting the same thing again

**Comment by homuler** [comment]
>>    It's because the arity is different...
>>
>>    graph.ObserveOutputStream("hand_landmarks", HandLandmarksCallback).AssertOk();
>
> Yes I am using it like this

???
https://github.com/homuler/MediaPipeUnityPlugin/issues/43#issuecomment-772326221
Then I don't understand what the above comment means.

> And getting the same thing again

Without knowing the exact code, I can't help you.
`unexpected EOF` occurs usually when the process is killed or aborted (Gog did not finish writing its buffer to the log file), and probably the root cause does not appear in the shared log.

**Comment by elblogbruno** [comment]
I will investigate what the problem could be furthermore. Maybe it is related because I am using Unity 2020.2.1?

**Comment by elblogbruno** [comment]
```
public override Status StartRun(Texture texture) {
    Debug.Log("This graph is for testing official examples. You can customize the graph by editing `official_demo_android.txt` (default is `hand_tracking_mobile.pbtxt`)");

    graph.ObserveOutputStream(handLandmarksStream, HandLandmarksCallback).AssertOk();
    
    sidePacket = new SidePacket();
    sidePacket.Emplace("num_hands", new IntPacket(2));

    destinationNativeTexturePtr = texture.GetNativeTexturePtr();
    destinationWidth = texture.width;
    destinationHeight = texture.height;

    gpuHelper.RunInGlContext(BuildDestination).AssertOk();
    sidePacket.Emplace(destinationBufferName, outputPacket);

    
    stopwatch.Start();
    Debug.LogWarning("HAND TRACKING GRAPH START RUN");
    
    return graph.StartRun(sidePacket);
  }
  
  [AOT.MonoPInvokeCallback(typeof(CalculatorGraph.NativePacketCallback))]
  static IntPtr HandLandmarksCallback(IntPtr ptr) {
    using (var packet = new NormalizedLandmarkListVectorPacket(ptr, false))
    {
      List<NormalizedLandmarkList> landmarks = packet.Get();
      Debug.Log("CALLBACK ");
      Debug.Log(landmarks[0]);
      return Status.Ok().mpPtr;
    }
  }
```
Or maybe there is something wrong here 🤨

**Comment by homuler** [comment]
This code won't work, because `landmarks` can be empty.
```cs
Debug.Log(landmarks[0]);
```
And note that `HandLandmarksCallback` **must not** throw exceptions, because native code can't handle it.

**Comment by elblogbruno** [comment]
> This code won't work, because `landmarks` can be empty.
> 
> ```cs
> Debug.Log(landmarks[0]);
> ```
> 
> And note that `HandLandmarksCallback` **must not** throw exceptions, because native code can't handle it.

so maybe it is that because it is null it throws an exception that blocks the thread?

**Comment by homuler** [comment]
I don't know. I think the behavior is not defined.
At least you need to handle exceptions.
```cs
static IntPtr HandleLandmarksCallback(IntPtr packetPtr) {
  try {
    using (var packet = new NormalizedLandmarkListVectorPacket(ptr, false)) {
      List<NormalizedLandmarkList> landmarks = packet.Get();
      Debug.Log("CALLBACK ");
      Debug.Log(landmarks[0]);
      return Status.Ok().mpPtr;
    }
  } catch (Exception e) {
    return Status.FailedPrecondition(e.ToString());
  }
}
```

**Comment by elblogbruno** [comment]
hmm ok thanks

**Comment by elblogbruno** [comment]
Still can't get to read data from the graph. I have tried both synchronously and asynchronously.

**Comment by elblogbruno** [comment]
I will try latest commit of repo

**Comment by holaworlds** [comment]
Hi @elblogbruno  could you figure out this? I am confused about how to assign the output to a variable. I am using `ObjectDetection3dGraph` example but concept is same as I see. 

should'nt we have a return value from HandLandmarksCallback function so that we can assign it to a variable and use it? HandleLandmarksCallback returns only status. 

@homuler  if you have any idea can you share it?
